src/Datadog.Trace.ClrProfiler.Managed/Integrations/GraphQLIntegration.cs

```diff
@@ -0,0 +1,351 @@
+
+            try
+            {
+                scope = tracer.StartActive(ValidateOperationName, serviceName: serviceName);
+                var span = scope.Span;
+                span.Type = SpanTypes.GraphQL;
+
+                span.SetTag(Tags.GraphQLSource, source);
+
+                // set analytics sample rate if enabled
+                var analyticsSampleRate = tracer.Settings.GetIntegrationAnalyticsSampleRate(IntegrationName, enabledWithGlobalSetting: false);
+                span.SetMetric(Tags.Analytics, analyticsSampleRate);
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorException("Error creating or populating scope.", ex);
+            }
+
+            return scope;
+        }
+
+        private static Scope CreateScopeFromExecuteAsync(object executionContext)
+        {
+            if (!Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName))
+            {
+                // integration disabled, don't create a scope, skip this trace
+                return null;
+            }
+
+            Tracer tracer = Tracer.Instance;
+            string source = executionContext.GetProperty("Document")
+                                            .GetProperty<string>("OriginalQuery")
+                                            .GetValueOrDefault();
+            string operationName = executionContext.GetProperty("Operation")
+                                                   .GetProperty<string>("Name")
+                                                   .GetValueOrDefault();
+            string operationType = executionContext.GetProperty("Operation")
+                                                   .GetProperty("OperationType")
+                                                   .ToString();
+            string serviceName = string.Join("-", tracer.DefaultServiceName, ServiceName);
+
+            Scope scope = null;
+
+            try
+            {
+                scope = tracer.StartActive(ExecuteOperationName, serviceName: serviceName);
+                var span = scope.Span;
+                span.Type = SpanTypes.GraphQL;
+                span.ResourceName = $"{operationType} {operationName ?? "operation"}";
+
+                span.SetTag(Tags.GraphQLSource, source);
+                span.SetTag(Tags.GraphQLOperationName, operationName);
+                span.SetTag(Tags.GraphQLOperationType, operationType);
+
+                // set analytics sample rate if enabled
+                var analyticsSampleRate = tracer.Settings.GetIntegrationAnalyticsSampleRate(IntegrationName, enabledWithGlobalSetting: false);
+                span.SetMetric(Tags.Analytics, analyticsSampleRate);
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorException("Error creating or populating scope.", ex);
+            }
+
+            return scope;
+        }
+
+        private static string ConstructErrorMessage(dynamic errors)
```

Comment:
What type are the errors?

Reply:
The errors are of type [ExecutionErrors](https://github.com/graphql-dotnet/graphql-dotnet/blob/cfc764b5adf95e36ef0048146a032285f7c0e53f/src/GraphQL/Execution/ExecutionErrors.cs). I resorted to using a dynamic type here because had trouble working with its public interface to enumerate things correctly. The type itself is an `IEnumerable<ExecutionError>` (but that makes `foreach` loops difficult to use since it's a specific generic type) and I don't know how to use the indexer without the compile-time reference.

Reply:
Would `GetEnumerator()` work?

Reply:
It also may be worth wrapping this method body in a try catch, as it's one of the more complicated parts of this integration. That way if this parsing fails for whatever reason, they don't lose the whole span.

Reply:
Good ideas Colin. Both of these suggestions have been implemented in ee035cd8adcb0cada6cbc083e291ed12cebf3c13 and b67b67667ad26a3555f48b720430eeb76d814b21.